File: dev_ws/src/learning_parameter_python/learning_parameter_python/param.py
"""
 @Description: 
 @version: v1.0
 @Author: HTY
 @Date: 2022-07-25 00:55:27
"""

import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import ParameterDescriptor


class Param(Node):
    def __init__(self, node):
        super(Param, self).__init__(node)
        self.timer = self.create_timer(2, self.timer_callback)
        param_description = ParameterDescriptor(description="This param is for learning.")
        self.declare_parameter('learn_param', 'world', param_description)

    def timer_callback(self):
        param = self.get_parameter('learn_param').get_parameter_value().string_value
        self.get_logger().info("Hello %s!" % param)
        new_param = rclpy.parameter.Parameter('learn_param', rclpy.Parameter.Type.STRING, param)
        all_new_params = [new_param]
        self.set_parameters(all_new_params)


def main(args=None):
    rclpy.init(args=args)
    param_node = Param("param_node")
    rclpy.spin(param_node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: dev_ws/src/learning_parameter_cpp/src/param.cpp
/*
 * @Description: 
 * @version: v1.0
 * @Author: HTY
 * @Date: 2022-07-25 14:54:12
 */

#include "rclcpp/rclcpp.hpp"

class Param : public rclcpp::Node
{
public:
    explicit Param(const std::string & node) : Node(node)
    {
        this->declare_parameter<std::string>("learn_param", "world");
        timer = this->create_wall_timer(std::chrono::seconds(1), std::bind(&Param::respond, this));
    }
    void respond()
    {
        this->get_parameter("learn_param", param_string);
        RCLCPP_INFO(this->get_logger(), "Hello %s", param_string.c_str());
    }

private:
    std::string param_string;
    rclcpp::TimerBase::SharedPtr timer;
};


int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Param>("param_node"));
    rclcpp::shutdown();
    return 0;
}
